elastic/beats
# Custom python script to insert user-supplied root CAs and mount points

## What does this PR do? 

This is probably one of the more...creative things I've done. This is my second attempt at creating some kind of script that would allow the users to point the plugin at a custom root cert. From what I can tell, no other plugins really attempt to solve this problem. This particular solution meets a few requirements I had in mind:

- Relatively small build-time requirements for the user
- Doesn't require us to host any additional build artifacts


This is a (relatively) small python script that uses the Docker Registry API to find the plugin, manually download the raw plugin artifacts, and build the plugin. This is actually rather similar to what happens under the hood when you run `docker plugin install`, we just do it ourselves so we can insert user config. Despite the fact that docker plugins and "normal" containers are stored in the Registry in a nearly identical way, we can't run `docker pull` on a plugin.

From the docs:

You can also specify a custom name and version tag for the rebuild plugin via `--tag` and `--name`. By default, the rebuilt plugin will iterate the patch release of the latest plugin.
By defualt, this script will build a new plugin based on the latest release available from the docker store. To specify a different version, use the `--from_release` argument.

This script rebuilds a plugin and adds a custom bindmount path. This can be used to add user-supplied certificates to a Plugin via a mounted directory. 
The host directory will be mounted as `/usr/local/share/ca-certificates` inside the plugin.

After the bindmount has been created, you can adjust the bindmount source and destination via `docker plugin set`:

`$ docker plugin set elastic/elastic-logging-plugin:8.0.0 CUSTOM_DIR.source="/new/source/bindmount"`

In cases where you're building the plugin on a system that doesn't have access to the docker daemon where you want to run the plugin, 
you can either use `docker plugin push` to push to your own docker repo (note that you must change the name with `--name`), or use the `--to_tar` flag with a target directory to generate a tarball.
For example: `./setmount.py --to_tar /path/to/write/tar/new_plugin.tar.gz /path/to/certs/on/docker/host`
To generate a docker plugin from the tarball, unpack the `.tar.gz` archive, and then run `docker plugin create NEW_PLUGIN_NAME PATH_TO_UNPACKED_DIRECTORY`.



My python is a tad rusty, so forgive me if there's a few stylistic errors here.

## Why is it important?

Many users will have Elastic Stack setups that require a custom cert.

## Checklist


## Author's Checklist

- [ ] Add user-facing docs on usage

## How to test this PR locally

- pull down PR, `cd` to `tools` directory.
- Install the `requests` library if you don't have it already.
- run `./setmount.py /path/to/certs/on/docker/host`
- run `./setmount.py --to_tar /path/to/certs/on/docker/host` to test tar writing
- run `docker plugin inspect ...` to check for the presence of the mount specified above
- enable or install the plugin, run a container with the plugin.
